### PR TITLE
Restore full Cygwin64 on Jenkins

### PR DESCRIPTION
**Pull request by dra27**
#9927 restores shared library and natdynlink support for Cygwin64. We no longer need to run Cygwin64 with `--disable-shared` - indeed, it's better not to (in order to catch errors in any new runtime functions which should be being exported).

This can't be merged yet, since it requires:
- [x] alainfrisch/flexdll#89 to be merged and a new release cut
- [x] New release of flexlink - [0.39 released](https://github.com/alainfrisch/flexdll/releases/tag/0.39)
- [x] Cygwin64's flexdll package to be updated to this new release - [0.39 available as a test release](https://cygwin.com/packages/summary/flexdll.html)
- [x] Inria Cygwin64 worker Cygwin64 installation to be updated with that new package

### CODE REVIEW

**Comment by dra27** [comment]
@shindere / @xavierleroy - would you like to update both the Cygwin workers and Jenkins to have the FlexDLL 0.39 package from Cygwin? It needs setup to be run manually so that "Test" packages can be checked. It should then automatically offer 0.39-1 as an update to the flexdll package (if it doesn't, you might need a different mirror - https://mirrorservice.org has definitely updated).

I've tested it on my machine, but it would of course be good to have trunk and 4.12 being regularly tested with a full Cygwin64 build.

**Comment by xavierleroy** [comment]
I updated the Cygwin64 worker with FlexDLL 0.39-1.  Maybe we can keep the old FlexDLL on the Cygwin32 worker for a bit longer?  As to "updating Jenkins to have FlexDLL" I don't know what that means.

**Review by xavierleroy — APPROVED** [review]
This looks good to me, and now we have ways to test it!  Thank you very much.

**Comment by dra27** [comment]
Sorry, I don’t know if it was “damn you autocorrect” or just Friday madness but I meant *on* Jenkins, not *and* Jenkins!

**Comment by xavierleroy** [comment]
So, Jenkins ran the testsuite on Cygwin64 and there's a handful of errors like
```
>      14 [main] ocamlrun 10122 child_info_fork::abort: address space needed by 'dllunix.so' (0x440000) is already occupied
> Fatal error: exception Unix.Unix_error(Unix.EAGAIN, "create_process", "/bin/sh")
```

It also happens when calling `Unix.fork()`.  

I'm afraid this is not the end of the Cygwin64 story...

**Comment by dra27** [comment]
That's strange, given that the DLL should never have a base address that low. I'll have a look - we're at least now in the realm of Cygwin64 not working, rather than FlexDLL!

**Comment by xavierleroy** [comment]
I may have botched the Cygwin upgrade (or upgraded the wrong machine...).  Let me try again.

**Comment by xavierleroy** [comment]
After a proper upgrade of the Cygwin64 host, CI succeds!  Happy end for a long story.  Thanks for all the efforts.
